### ISSUE TITLE
Implement authorization for file download URLs

### ISSUE THREAD

**Issue by aleksihakli** [benoitbryon/django-downloadview#138]
Hi,

_tl; dr: Would this project benefit from the ability to sign download URLs (cryptographically and with expiration)?_

I thought I would open a discussion on adding download URL signatures.

I recently implemented cryptographic authorization on top of URLs that were generated directly by the storage backend, much like with S3.

These were served with nginx by using X-Accel and a custom view that generated serve requests to the proxy server, offloading the file serving from Django.

The idea is fairly simple and I think many people could benefit from it. Implementation just requires

- a Storage class mixin for specializing URL generation to add signatures in query parameters, and;
- a decorator that validates file download URLs for the download views.

The best thing is that download views will work with or without the signature.

Following a naive example of the idea of the implementation. Please bear in mind that these examples are untested and would, of course, need to be further adapted for `django_downloadview`. 

```python
# demoproject/download/example.py

from django.conf import settings
from django.core.files.storage import Storage
from django.core.signing import TimestampSigner, SignatureExpired, BadSignature

signer = TimestampSigner()
expiration = getattr(settings, 'SIGNED_URL_EXPIRATION_SECONDS', 0)


class SignedURLMixin(Storage):
    """ Mixin for generating signed file URLs with storage backends. Add X-Signature query parameters to the normal URLs generated by the backend"""

    def url(self, name):
        path = super(SignedURLMixin, self).url(name)
        signature = signer.sign(path)
        return '{}?X-Signature={}'.format(path, signature)


class SignedFileSystemStorage(SignedURLMixin, FileSystemStorage):
    pass

def signature_required(function):
    """ Decorator that checks for X-Signature query parameter to authorize specific user access. """

    def wrap(request, *args, **kwargs):
        try:
            signature = request.GET.get('X-Signature')
            path = signer.unsign(signature, max_age=expiration)
        except BadSignature:
            return HttpResponseForbidden('Signature invalid')
        except SignatureExpired:
            return HttpResponseForbidden('Signature expired')
        except Exception:
            return HttpResponseForbidden('Signature error')
        else:
            if path != request.path:
                return HttpResponseForbidden('Signature mismatch')

        return function(request, *args, **kwargs)

    wrap.__doc__ = function.__doc__
    wrap.__name__ = function.__name__
    return wrap
```

Then the usage can simply be:

```python
# demoproject/download/urls.py

# Django is set up with
# DEFAULT_FILE_STORAGE='demoproject.download.example.SignedStorage'

from django.conf.urls import url, url_patterns
from django_downloadview import ObjectDownloadView
from demoproject.download.models import Document  # A model with a FileField
from demoproject.download.example import signature_required

# ObjectDownloadView inherits from django.views.generic.BaseDetailView.
download = ObjectDownloadView.as_view(model=Document, file_field='file')

url_patterns = ('',
    url('^download/(?P<slug>[A-Za-z0-9_-]+)/$', signature_required(download), name='download'),
)

# demoproject/download/template.html
# URLs in templates are generated with the storage class URL implementation

<a href="{{ object.file.url  }}">Click here to download.</a>
```

The S3 Boto storage backend uses a similar approach and makes it possible to generate URLs in user templates and then authorize S3 access with those URLs. This vanilla Django approach makes it very easy to emulate that behaviour.

Additional hardening can then be achieved with:

- Adding random salts to signing, and expiration times to the [TimestampSigner](https://docs.djangoproject.com/en/1.10/topics/signing/#django.core.signing.TimestampSigner)
- Only ever using signed download links generated with the storage backend using `{{ file.url }}` 

This approach only lacks in that it introduces non-cacheable URLs that require slight computation to decrypt.

Inspiration was received from Grok. You can find more information on generic URL signatures in his weblog:

- http://grokcode.com/819/one-click-unsubscribes-for-django-apps/

If signatures are appended to URLs with existing query parameters, a more sophisticated solution has to be used. For example:

- https://stackoverflow.com/questions/2506379/add-params-to-given-url-in-python

**Comment by aleksihakli** [comment]
Hi @benoitbryon and @Natim, would this be something that is needed in this project? I could implement this in a PR but would like to know if there has been any previous discussion or opinions on the subject.

**Comment by Natim** [comment]
If you need it then you should probably start a PR.

**Comment by Natim** [comment]
@aleksihakli feel free to do so if you think it is still interesting yes.

**Comment by aleksihakli** [comment]
I'll try and find time for making a PR. The example code should work as-is. I've tested it locally in the past. It could be added to the codebase with a few unit tests.

**Comment by aleksihakli** [comment]
@Natim the docs seem to be of the old version by the way, is it possible to update the RTD site with the new contents which also describe using the new configuration options?

**Comment by Natim** [comment]
Yes I asked @benoitbryon about that. But we might consider asking someone from RTD to give us access.

**Comment by benoitbryon** [comment]
Hi there :)
Natim, I just granted you maintainer of django-downloadview on RTD. Is it ok for you?
Feel free to make the changes you think are necessary.

**Comment by Natim** [comment]
Thank you Benoît I will have a look

Le mar. 14 janv. 2020 à 23:05, Benoît Bryon <notifications@github.com> a
écrit :
